handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,223 @@
+        if (timeout <= 0) {
+            timeoutNanos = 0;
+        } else {
+            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
+        }
+        this.closeOnTimeout = closeOnTimeout;
+        this.suppressReadAfterTimeout = suppressReadAfterTimeout;
+    }
+
+    /**
+     * Creates a new instance with {@code closeOnTimeout == true} and {@code ignoreReadAfterTimeout == true}
+     *
+     * @param timeout
+     *        read timeout
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit) {
+        this(timeout, unit, true, true);
+    }
+
+    @Override
+    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+        removed = false;
+
+        if (channelOnWriteListener == null) {
+            channelOnWriteListener = new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    //ensure that only the last write listener in a batch would setup one final timeout task
+                    //or nothing in a case of error
+                    disable();
+                    timedOut = false;
+                    if (future.isSuccess()) {
+                        if (removed) {
+                            return;
+                        }
+                        writeFinishTime = System.nanoTime();
+                        //init readFinishTime to writeFinishTime - readFinishTime > 0 be true if we don't get response
+                        readFinishTime = writeFinishTime - 1;
+                        if (timeoutNanos > 0) {
+                            timeout = ctx.executor().schedule(readTimeoutTask, timeoutNanos, TimeUnit.NANOSECONDS);
+                        }
+                    }
+                }
+            };
+        }
+
+        if (readTimeoutTask == null) {
+            readTimeoutTask = new ReadTimeoutTask(ctx);
+        }
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        readFinishTime = System.nanoTime();
+        if (timeoutShouldBeCancelled(ctx, msg)) {
+            disable();
+        }
+        if (suppressReadAfterTimeout && timedOut) {
+            ctx.fireUserEventTriggered(ReadSuppressed.INSTANCE);
+            ReferenceCountUtil.release(msg);
+        } else {
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelOnWriteListener));
```

Comment:
Semantic of this handler:
You write a client and you need to make classical timeout semantic. You wrote everything successfully to a socket and want to get a result in 100ms from the server and pass it to the caller, or return "204" entity to the caller if no answer got in this timeout.

So, this handler doesn't cover situation with "write doesn't complete". It's to handle fair server response time between finishing write on the client side and reading a response. This is how it should work in my mind. Do you propose to extend semantic?